### ISSUE TITLE
Documentation update regarding the safety of `Sys.opaque_identity` with respect to garbage collection

### ISSUE THREAD

**Issue by Fizzixnerd** [ocaml/ocaml#13287]
Following discussion in https://github.com/yallop/ocaml-ctypes/issues/571#issuecomment-2220520086 , I am wondering if the following assumption about `Sys.opaque_identity` is true:

> The argument to `Sys.opaque_identity` cannot be collected as garbage.

The reason I ask is detailed in the above discussion, but briefly we are looking for a way to ensure that particular memory locations are not garbage collected until _at least_ a specific point in a program when using `Ctypes`.  For example, we would like to ensure the following code is safe:

```ocaml
let strchr = Foreign.foreign "strchr" (ptr char @-> char @-> returning (ptr char)) in
let p = CArray.of_string "abc" in
let q =  strchr (CArray.start p) 'a' in
let () = Gc.compact () in
let () = ignore (Sys.opaque_identity (p, q)) in (**** This is the important line ****)
Printf.printf "%c\n" !@q
```

There remains some question if it is, and perhaps the documentation should be updated with information in whichever case it is, as it seems important and the folk-lore is not established in the community.

I tried to check the source, but it's in `sys.ml.in` and I couldn't trace it further back than that.

CC @yallop @edwintorok for interest.

**Comment by lthls** [comment]
The argument of `Sys.opaque_identity` cannot be collected *before* the call to `Sys.opaque_identity`, but can be collected *after* the call.
You can use `Callback.register_named_value` if you want to make sure a value is kept alive forever, or store it in a global structure.

**Comment by Fizzixnerd** [comment]
Thanks.  Would you be open to a PR modifying the documentation to mention this fact?

**Comment by lthls** [comment]
You can open a PR, of course, but I'm not sure why the existing documentation is not enough. Perhaps adding your use case (keeping values alive until a certain point) as an additional example ?
Note that in your code above, I consider that `p` may have been collected between the call to `Sys.opaque_identity` and the call to `!@q`, even though you will not observe it in practice. If you want to make sure that `p` is alive across the call to `!@q`, you need to put the call to `Sys.opaque_identity` after the expression.

**Comment by Fizzixnerd** [comment]
You are indeed correct that my code contained a bug, even with the opaque_identity!

Regarding the existing documentation: it contains the following words, which I think are the source of the confusion:

> For the purposes of optimization, opaque_identity behaves like an unknown (and thus possibly side-effecting) function.
> 
> **At runtime, opaque_identity disappears altogether.**

(emphasis mine).  This has, it seems, (see https://github.com/yallop/ocaml-ctypes/issues/571, https://github.com/xapi-project/xen-api/pull/3669 as evidence) led people to believe that this technique would not be effective at preventing the _runtime_ garbage collector from collecting it, even if the _optimizer_ didn't optimize it out.

**Comment by lthls** [comment]
Well, initially it only prevented parts of the compiler from optimising around it, and disappeared before liveness analysis. This meant that it didn't work for keeping values alive. This led to #9412, which made it stricter (in practice pushing it down to the end of the compiler's passes), and enforced the property of keeping arguments alive at runtime.

**Comment by xavierleroy** [comment]
Addressed in #13287.
